[shard] add ShardedTensor.cpu()

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74941



Add ShardedTensor.cpu() API, we choose to add this first instead of `ShardedTensor.to()` bc the latter one is ambiguous and tricker (i.e. we need to figure out what's the meaning of ShardedTensor.to("cuda")).

Note that when moving ShardedTensor to CPU from GPU, user need to explicitly pass in a new process_group that is compatible with the device, otherwise we will error out.

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133/)

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133)

Commit:
[shard] add ShardedTensor.to

Add ShardedTensor.to API.
This API now only accepts `device`, we will add `dtype` and other flags https://pytorch.org/docs/stable/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to

Note that when passing device that is different from the current sharded tensor (i.e. sharded tensor local shards on GPU but we want to move it to CPU), user need to explicitly pass in a new process_group that is compatible with the device, otherwise we will error out.

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -313,6 +313,30 @@ def gather(
 
                     out_narrow_view.copy_(tensor)
 
+    def to(self,
+           device=None,
+           process_group=None):
```
Don't we need docs for this method?

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -313,6 +313,30 @@ def gather(
+           device=None,
+           process_group=None):
+        if device is not None:
+            device_is_cuda = "cuda" in str(device)
```
nit:

```
dev = torch.device(device) # this gives us free validation of device
device_is_cuda = dev.type == 'cuda'
```

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -313,6 +313,30 @@ def gather(
+            device_is_cuda = "cuda" in str(device)
+            pg_is_nccl = isinstance(self._process_group, distributed_c10d.ProcessGroupNCCL)
+
+            if device_is_cuda != pg_is_nccl:
+                assert process_group is not None, "need to pass in process_group as argument!"
```
Hmm so I don't think we should have device specific or pg specific logic like this. We should clearly document the function mentioning that it is the users responsibility to ensure that the PG passed it works for the device passed in.

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1464,6 +1465,61 @@ def test_gather_uneven(self) -> None:
+        st.to(device="cpu", process_group=gloo_pg)
+        spec = st.sharding_spec()
+        metas = st.metadata().shards_metadata
+        self.assertIsInstance(spec, ChunkShardingSpec)
```
Can we validate that the spec is the same as passed in during tensor creation except that the device is now cpu?

Commit:
Update on "[shard] add ShardedTensor.to"


Add ShardedTensor.to API.
This API now only accepts `device`, we will add `dtype` and other flags https://pytorch.org/docs/stable/generated/torch.Tensor.to.html?highlight=#torch.Tensor.to

Note that when passing device that is different from the current sharded tensor (i.e. sharded tensor local shards on GPU but we want to move it to CPU), user need to explicitly pass in a new process_group that is compatible with the device, otherwise we will error out.

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133/)

[ghstack-poisoned]

Commit:
Update on "[shard] add ShardedTensor.cpu()"




Add ShardedTensor.cpu() API, we choose to add this first instead of `ShardedTensor.to()` bc the latter one is ambiguous and tricker (i.e. we need to figure out what's the meaning of ShardedTensor.to("cuda")).

Note that when moving ShardedTensor to CPU from GPU, user need to explicitly pass in a new process_group that is compatible with the device, otherwise we will error out.

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133/)

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -313,6 +314,63 @@ def gather(
+            it is the user's responsiblity to explicitly pass in a new process_group that
+            is compatible with CPU.
+        """
+        # TODO: make this a __torch_function__ op once ShardedTensor becomes a
+        # torch.Tensor subclass
```
nit: filing a gh issue for this will ensure we don't lose track of it :)

pritamdamania87 commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -313,6 +314,63 @@ def gather(
+            raise RuntimeError("Only `torch.contiguous_format` or "
+                               "`torch.preserve_format` is supported!")
+        all_on_cpu = True
+        for meta in self.metadata().shards_metadata:
+            if meta.placement.device().type != "cpu":
```
nit:

```
all_on_cpu = True
for meta in self.metadata().shards_metadata: 
  all_on_cpu &= (meta.placement.device().type != "cpu")
```

pritamdamania87 commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1464,6 +1465,61 @@ def test_gather_uneven(self) -> None:
+        spec = ChunkShardingSpec(
+            dim=0,
+            placements=[
+                "rank:0/cuda:0",
+                "rank:1/cuda:1",
+                "rank:2/cuda:2",
+                "rank:3/cuda:3",
```
Maybe we should also test a spec where some shards are on cpu and some on GPU and then we call `.cpu()`.

wanchaol replied:
hmmm how we should represent such a ShardedTensor, is it being managed by one process group? Maybe a processgroupGloo can manage a ShardedTensor where local shards can be on CPU/GPU together?

wanchaol replied:
added.

pritamdamania87 reviewed (approved):
LGTM! Lets add one more additional test as mentioned inline.

Commit:
Update on "[shard] add ShardedTensor.cpu()"




Add ShardedTensor.cpu() API, we choose to add this first instead of `ShardedTensor.to()` bc the latter one is ambiguous and tricker (i.e. we need to figure out what's the meaning of ShardedTensor.to("cuda")).

Note that when moving ShardedTensor to CPU from GPU, user need to explicitly pass in a new process_group that is compatible with the device, otherwise we will error out.

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133/)

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133)

[ghstack-poisoned]

Commit:
Update on "[shard] add ShardedTensor.cpu()"




Add ShardedTensor.cpu() API, we choose to add this first instead of `ShardedTensor.to()` bc the latter one is ambiguous and tricker (i.e. we need to figure out what's the meaning of ShardedTensor.to("cuda")).

Note that when moving ShardedTensor to CPU from GPU, user need to explicitly pass in a new process_group that is compatible with the device, otherwise we will error out.

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133/)

Differential Revision: [D35238133](https://our.internmc.facebook.com/intern/diff/D35238133)

[ghstack-poisoned]
